### PR TITLE
bpo-37531: Fix regrtest timeout for subprocesses

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
Fix regrtest timeout for subprocesses.

Naming things is the hardest thing in computer science. Interested in any reviews in a better name If any for the new function I introduced. cc @vstinner 

https://bugs.python.org/issue37531

### CODE REVIEW

**Comment by ammaraskar** [comment]
Hey @nanjekyejoannah, did you accidentally attach the wrong bpo ticket for this? 3753 seems really old and unrelated.

**Comment by ZackerySpytz** [comment]
It seems that you have changed the file mode for several files under `Lib/test`. Why? I don't think those changes should be made.

**Comment by ZackerySpytz** [comment]
It seems that you have changed the file mode of several files under `Lib/test`. Why? I don't think those changes should be made.

File: Lib/test/libregrtest/runtest.py [hunk 1]
```diff
@@ -35,6 +36,7 @@
     INTERRUPTED: '%s interrupted',
     CHILD_ERROR: '%s crashed',
     TEST_DID_NOT_RUN: '%s run no tests',
+    TIMEOUT: '%s test did not finish',
```
Comment (vstinner) [inline Lib/test/libregrtest/runtest.py]: I suggest:

```suggestion
    TIMEOUT: '%s timed out',
```

File: Lib/test/libregrtest/runtest.py [hunk 2]
```diff
@@ -179,6 +181,7 @@ def runtest(ns, test_name):
         FAILED           test failed
         PASSED           test passed
         EMPTY_TEST_SUITE test ran no subtests.
+        TIMEOUT         test failed due to timeout
```
Comment (vstinner) [inline Lib/test/libregrtest/runtest.py]:
```suggestion
         TIMEOUT         test timed out 
```

File: Misc/NEWS.d/next/Library/2019-07-09-19-38-26.bpo-3753.GX7s8S.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix `regrtest` timeout for subprocesses.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2019-07-09-19-38-26.bpo-3753.GX7s8S.rst]: You should better explain the change. The change only impacts "python3 -m test -jN --timeout=TIMEOUT", it ensures that a worker process is stopped with a timeout if it runs longer than TIMEOUT seconds.

**Review by vstinner — COMMENTED** [review]
It seems like you modified Lib/test/test_os.py permissions somehow, please revert this change.

**Comment by nanjekyejoannah** [comment]
I have made the requested changes; please review again.

**Comment by nanjekyejoannah** [comment]
> 
> 
> Hey @nanjekyejoannah, did you accidentally attach the wrong bpo ticket for this? 3753 seems really old and unrelated.

Yes, thanks for catching this. I made the necessary changes.

**Comment by nanjekyejoannah** [comment]
> 
> 
> It seems that you have changed the file mode of several files under `Lib/test`. Why? I don't think those changes should be made.

oh yes, I have reverted this.

**Comment by vstinner** [comment]
Overall, the change is great and seems to work as expected, but I tested it and it doesn't work as I expected:

* it doesn't report stdout/stderr on timeout
* regression: it ignores stdout/stderr on "CHILD_ERROR" whereas it was properly passed to the main process previously
* faulthandler  and Popen.communicate() use the same timeout, so it's unclear which one which trigger first: I tuned timeouts to give more time to faulthandler. I prefer faulthandler since it dumps the traceback where Python was stuck, whereas Popen.communicate() timeout only returns "truncated" output with no traceback
* the timeout is reported as an exception, not as as "result"

I tried to add a commit to this PR but I wasn't allowed, so I forked this PR and created PR #15072 to add my changes on top of it.

**Comment by nanjekyejoannah** [comment]
Closing this then in favor of PR #15072

**Pull request closed without merging**